### DISCUSSION TITLE
How to bypass max_rows_to_read config?

### DISCUSSION THREAD

**Discussion by brunomrpx** [ClickHouse/ClickHouse#53646] [Q&A]
Hi guys!

I'm working on a ClickHouse cluster that stores time-series data for an analytics API. The `max_rows_to_read` config is set to `3bi`; for some large queries, we still receive an exception that this limit has been reached. We're trying to avoid increasing this number to not overload the servers. We already tried to make the API "split" the query into smaller queries using pagination (LIMIT and OFFSET) and search a small time interval, but the limit is stills being reached. Are there any optimizations that can be done to don't reach this limit? What are the harms of increasing this config?

Following is an example of how to simulate the behavior. 

- Table definition
```sql
CREATE TABLE test_clk_limits.logs
(
   `time` DateTime CODEC(DoubleDelta, LZ4),
   `server` String,
   `client` String,
   `metric_1` UInt64,
   `metric_2` UInt64,
   `metric_3` UInt64,
   `metric_4` String
)
ENGINE = MergeTree()
PRIMARY KEY (client, metric_4, time)
ORDER BY (client, metric_4, time);
```

- Insert sample data
```sql
INSERT INTO test_clk_limits.logs
   SELECT 
      toDateTime('2023-01-01T00:00:00') + INTERVAL number MINUTE as time,
      'server-1' AS server,
      'xxxx' as client,
      rand() as metric_1,
      rand() as metric_2,
      rand() as metric_3,
      toString(rand()) as metric_4
   FROM numbers(24 * 60 * 30 * 3); -- insert data for 3 months
```

- Search for a given time range period (overriding the `max_rows_to_read` to reach the limit)
```sql
SELECT
    time,
    metric_1,
    metric_2
FROM test_clk_limits.logs
WHERE
    client = 'xxxx'
    AND time BETWEEN '2023-01-01T00:00:00' AND '2023-02-01T00:00:00'
SETTINGS max_rows_to_read=100000;
```

- Result:
```sql
Code: 158. DB::Exception: Received from 127.0.0.1:9441. DB::Exception: Limit for rows (controlled by 'max_rows_to_read' setting) exceeded, max rows: 100.00 thousand, current rows: 129.60 thousand.
```

Any help will be very useful, thank you!

**Comment by UnamedRus** [comment]
> What are the harms of increasing this config?

It's (limit) not enabled/enforced by default.
You can just remove that setting

**Comment by mo-avatar** [comment]
Large outputs may lead to large IO Throughput， if your cluster have an high concurrency, it may influnence the query speed of other queries. But still you can try just set config to 0 to disable it and see what happens, or may be you can just open it  within your certain query with settings gramar and let the default config be it is.
